libraries/AP_Beacon/AP_Beacon_Nooploop.cpp

```diff
@@ -0,0 +1,238 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include <GCS_MAVLink/GCS.h>
+#include "AP_Beacon_Nooploop.h"
+#include <ctype.h>
+#include <stdio.h>
+
+#define NOOPLOOP_INVALID_VAL -8388000
+
+extern const AP_HAL::HAL& hal;
+
+AP_Beacon_Nooploop::AP_Beacon_Nooploop(AP_Beacon &frontend, AP_SerialManager &serial_manager) :
+    AP_Beacon_Backend(frontend)
+{
+    _uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Beacon, 0);
+    if (_uart != nullptr) {
+        _uart->begin(serial_manager.find_baudrate(AP_SerialManager::SerialProtocol_Beacon, 0));
+    }
+}
+
+// return true if sensor is basically healthy (we are receiving data)
+bool AP_Beacon_Nooploop::healthy()
+{
+    // healthy if we have parsed a message within the past 300ms
+    return ((AP_HAL::millis() - _last_update_ms) < AP_BEACON_TIMEOUT_MS);
+}
+
+// update the state of the sensor
+void AP_Beacon_Nooploop::update(void)
+{
+    // return immediately if not serial port
+    if (_uart == nullptr) {
+        return;
+    }
+
+    // check uart for any incoming messages
+    uint32_t nbytes = _uart->available();
```

Comment:
Can't do this; if you somehow get a very large number of bytes from the uart you might fall out of the sky.  Please use `MIN(...)` with a suitable value here.

Reply:
Thank you very much. I will fix this.

Reply:
I guess this MIN hasn't been added yet?  Maybe it's github playing tricks on me but it looks like this issue is still outstanding.

Reply:
May I ask a question about this? Is it really necessary? I search codes for example to do this. But I can not find any other driver use MIN for available().

Reply:
```
libraries/AP_Camera/AP_RunCam.cpp:719:    uint32_t avail = MIN(uart->available(), (uint32_t)RUNCAM_MAX_PACKET_SIZE);
libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:344:    uint32_t numc = MIN(_msp_port.uart->available(), 1024U);
libraries/AP_Proximity/AP_Proximity_LightWareSF45B.cpp:98:    uint32_t nbytes = MIN(_uart->available(), 1024U);
```

Those were the ones discoverable with a simple grep.

Reply:
Hi @peterbarker Thank you very much, I am very sorry that I miss it. It is my fault. I have fixed this

Reply:
In practice it would probably never cause a problem because the nooploop will likely behave itself.  Still, you know.. we have so many users and crazy stuff happens sometimes.  For example, one time a user connecte a serial port to an I2C port and the resulting noise crashes the autopilot.  So where possible, even if it's unlikely we add safety checks in there.